File: nix/packages/catls/catls.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Getopt::Long;
use File::Spec;
use File::Basename;
use Cwd 'abs_path';
use Carp;

# Default ignore patterns (regex strings)
my @ignore_regex = (
    qr/\.git\//,
    qr/\.svn\//,
    qr/\.hg\//,
    qr/__pycache__\//,
    qr/\.pytest_cache\//,
    qr/\.mypy_cache\//,
    qr/\.tox\//,
    qr/\.venv\//,
    qr/\.coverage/,
    qr/\.DS_Store/,
    qr/\.idea\//,
    qr/\.vscode\//,
    qr/.*_templ\.go$/,
    qr/LICENSE$/,
    qr/LICENSE\.md$/,
    qr/LICENSE\.txt$/,
);

# Default ignore directories (strings)
my @ignore_dir = (
    "node_modules", ".direnv", "build", "dist", "target",
    "venv", "env", ".env", "vendor", ".bundle", "coverage",
);

# Command-line options
my $show_all          = 0;
my $recursive         = 0;
my $content_pattern   = "";
my $show_line_numbers = 0;
my $debug             = 0;

# User-provided patterns
my @include_regex;     # strings or qr// for include
my @extra_ignore_regex;
my @extra_ignore_dir;

GetOptions(
    'all|a'           => \$show_all,
    'recursive|r'     => \$recursive,
    'ignore-regex=s@' => \@extra_ignore_regex,
    'ignore-dir=s@'   => \@extra_ignore_dir,
    'regex=s@'        => \@include_regex,
    'pattern=s'       => \$content_pattern,
    'line-numbers|n'  => \$show_line_numbers,
    'debug'           => \$debug,
) or croak "Error in command line arguments\n";

# Merge user-specified ignores
push @ignore_regex, map { qr/$_/ } @extra_ignore_regex;
push @ignore_dir,    @extra_ignore_dir;

# Positional args: directory and optional files
my $directory     = shift @ARGV // ".";
my @forced_files  = @ARGV;

# If forced files given, add exact-match include patterns
for my $f (@forced_files) {
    if (-f $f) {
        my $b = basename($f);
        push @include_regex, qr/^\Q$b\E$/;
    }
}

die "Error: '$directory' is not a valid directory\n"
    unless -d $directory;

log_debug("Ignore dirs: @ignore_dir");
log_debug("Ignore regex: @ignore_regex");
log_debug("Include regex: @include_regex");

# Collect files
my @all_files;
find_files($directory, \@all_files);

@all_files = sort @all_files;
unless (@all_files) {
    print "No files found in directory: $directory\n";
    exit 0;
}

# Output XML
print "<files>\n";
for my $file (@all_files) {
    my $rel = File::Spec->abs2rel($file, $directory);
    next unless should_include($rel, \@include_regex);
    if (should_ignore($rel, \@ignore_regex, \@ignore_dir)) {
        log_debug("Ignoring file: $rel");
        next;
    }
    my $safe_path = xml_escape($rel);
    print qq{<file path="$safe_path">\n};
    if (is_binary($file)) {
        print "  <binary>true</binary>\n";
        print "  <content>[Binary file - contents not displayed]</content>\n";
    } else {
        my $type = xml_escape(guess_filetype($file));
        print "  <type>$type</type>\n";
        if (open my $fh, "<:encoding(UTF-8)", $file) {
            my @lines    = <$fh>;
            close $fh;
            my @filtered = filter_content(\@lines, $content_pattern);
            my $total    = @lines;
            my $count    = @filtered;
            my $limit    = 100;
            my $need_truncate = (!$content_pattern && $count > 1000);
            @filtered = @filtered[0..$limit-1] if $need_truncate;
            print "  <content>\n";
            for my $pair (@filtered) {
                my ($ln, $text) = @$pair;
                if ($show_line_numbers) {
                    printf("%4d| %s", $ln, $text);
                } else {
                    print $text;
                }
            }
            if ($need_truncate) {
                print "... (" . ($total - $limit) . " more lines)\n";
            }
            print "  </content>\n";
        } else {
            print "  <error>", xml_escape($!), "</error>\n";
        }
    }
    print "</file>\n";
}
print "</files>\n";

#------------------ Subroutines ------------------#

sub find_files {
    my ($dir, $files) = @_;
    opendir my $dh, $dir or do {
        warn "Cannot open '$dir': $!\n";
        return;
    };
    for my $entry (readdir $dh) {
        next if $entry eq '.' or $entry eq '..';
        next if !$show_all and $entry =~ /^\./;
        my $path = File::Spec->catfile($dir, $entry);
        if (-d $path) {
            if (should_ignore($path, \@ignore_regex, \@ignore_dir)) {
                log_debug("Ignoring directory: $path");
            } elsif ($recursive) {
                find_files($path, $files);
            }
        }
        elsif (-f $path) {
            push @$files, $path;
        }
    }
    closedir $dh;
}

sub wildcard_to_regex {
    my ($pat) = @_;
    $pat = quotemeta($pat);
    $pat =~ s/\\\*/.*/g;
    $pat =~ s/\\\?/./g;
    return $pat;
}

sub is_binary {
    my ($file) = @_;
    # Try 'file' command
    if (my $out = `file --brief --mime-type '$file' 2>/dev/null`) {
        return $out !~ m{^text/}i;
    }
    # Fallback: look for NUL bytes
    if (open my $fh, '<:raw', $file) {
        read($fh, my $buf, 1024);
        close $fh;
        return $buf =~ /\0/;
    }
    return 1;  # if we can't read it, assume binary
}

sub guess_filetype {
    my ($file) = @_;
    my %map = (
        sh      => 'bash', bash  => 'bash',
        py      => 'python',
        js      => 'javascript',
        html    => 'html',
        nix     => 'nix',
        css     => 'css',
        json    => 'json',
        md      => 'markdown',
        xml     => 'xml',
        c       => 'c',
        cpp     => 'cpp',
        h       => 'c',
        toml    => 'toml',
        hpp     => 'cpp',
        java    => 'java',
        rs      => 'rust',
        go      => 'go',
        rb      => 'ruby',
        php     => 'php',
        pl      => 'perl',
        sql     => 'sql',
        templ   => 'templ',
        yml     => 'yaml',
        yaml    => 'yaml',
    );
    my ($ext) = $file =~ /\.([^.]+)$/;
    return $map{lc $ext} // '';
}

sub should_include {
    my ($path, $patterns) = @_;
    return 1 unless @$patterns;
    my $base = basename($path);
    for my $pat (@$patterns) {
        if (ref $pat eq 'Regexp') {
            return 1 if $path =~ $pat;
        } else {
            if ($pat =~ /[\*\?]/) {
                my $re = qr/^@{[ wildcard_to_regex($pat) ) ]}$/;
                return 1 if $base =~ $re or $path =~ $re;
            } else {
                my $re = qr/$pat/;
                return 1 if $path =~ $re;
            }
        }
    }
    return 0;
}

sub should_ignore {
    my ($path, $pat_regex, $pat_dirs) = @_;
    my $real = get_real_path($path);
    # directory-name and path-based ignores
    my @parts = File::Spec->splitdir($path);
    for my $d (@$pat_dirs) {
        # ignore matching directory names
        if ($d !~ m{/} and grep { $_ eq $d } @parts) {
            return 1;
        }
        # ignore path prefixes
        if ($d =~ m{/}) {
            my $ignore_root = get_real_path($d);
            if (defined $ignore_root and index($real, $ignore_root) == 0) {
                return 1;
            }
        }
    }
    # regex-based ignores
    for my $re (@$pat_regex) {
        return 1 if $path =~ $re;
    }
    return 0;
}

sub get_real_path {
    my ($p) = @_;
    for my $cmd (["realpath", $p], ["readlink", "-f", $p]) {
        if (my $out = qx{$cmd->[0] '$cmd->[1]' 2>/dev/null}) {
            chomp $out;
            return $out if -e $out;
        }
    }
    return abs_path($p);
}

sub filter_content {
    my ($lines, $pattern) = @_;
    my @out;
    if ($pattern) {
        my $re = qr/@{[ wildcard_to_regex($pattern) ]}/;
        for my $i (0 .. $#$lines) {
            push @out, [ $i+1, $lines->[$i] ] if $lines->[$i] =~ $re;
        }
    } else {
        for my $i (0 .. $#$lines) {
            push @out, [ $i+1, $lines->[$i] ];
        }
    }
    return @out;
}

sub xml_escape {
    my ($s) = @_;
    return '' unless defined $s;
    $s =~ s/&/&amp;/g;
    $s =~ s/</&lt;/g;
    $s =~ s/>/&gt;/g;
    $s =~ s/"/&quot;/g;
    $s =~ s/'/&apos;/g;
    return $s;
}

sub log_debug {
    return unless $debug;
    warn "Debug: ", @_;
}

File: modules/programs/catls/catls.py
from __future__ import annotations
import argparse
import os
import re
import subprocess
import sys
from pathlib import Path
from html import escape
from dataclasses import dataclass, field
from typing import cast


@dataclass
class Args:
    show_all: bool = False
    recursive: bool = False
    debug: bool = False
    default_ignore_globs: list[str] = field(
        default_factory=lambda: [
            ".git/*",
            ".svn/*",
            ".hg/*",
            "__pycache__/*",
            ".pytest_cache/*",
            ".mypy_cache/*",
            ".tox/*",
            ".venv/*",
            ".coverage",
            ".DS_Store",
            ".idea/*",
            ".vscode/*",
            "*_templ.go",
            "LICENSE",
            "LICENSE.md",
            "LICENSE.txt",
        ]
    )
    ignore_dir: list[str] = field(
        default_factory=lambda: [
            "node_modules",
            ".direnv",
            "build",
            "dist",
            "target",
            "venv",
            "env",
            ".env",
            "vendor",
            ".bundle",
            "coverage",
            "static",
        ]
    )
    globs: list[str] = field(default_factory=list)
    ignore_globs: list[str] = field(default_factory=list)
    directory: str = "."
    files: list[str] = field(default_factory=list)
    content_pattern: str = ""
    show_line_numbers: bool = False
    omit_bins: bool = False


def wildcard_to_regex(pattern: str) -> str:
    """Convert shell-style wildcard pattern to regex pattern."""
    result = re.escape(pattern)
    result = result.replace(r"\*", ".*").replace(r"\?", ".")
    return result


def is_binary(file_path: str) -> bool:
    """Check if a file is binary using the 'file' command."""
    try:
        result = subprocess.run(
            ["file", file_path],
            capture_output=True,
            text=True,
        )
        return "text" not in result.stdout.lower()
    except (
        subprocess.SubprocessError,
        FileNotFoundError,
    ):
        try:
            with open(file_path, "rb") as f:
                chunk = f.read(1024)
                return b"\0" in chunk
        except Exception:
            return True


def guess_filetype(file_path: str) -> str:
    """Guess file type based on extension."""
    ext = Path(file_path).suffix.lower().lstrip(".")

    filetypes = {
        "sh": "bash",
        "bash": "bash",
        "rb": "ruby",
        "py": "python",
        "js": "javascript",
        "html": "html",
        "nix": "nix",
        "css": "css",
        "json": "json",
        "md": "markdown",
        "xml": "xml",
        "c": "c",
        "cpp": "cpp",
        "h": "c",
        "toml": "toml",
        "hpp": "cpp",
        "java": "java",
        "rs": "rust",
        "go": "go",
        "php": "php",
        "pl": "perl",
        "sql": "sql",
        "templ": "templ",
        "yml": "yaml",
        "yaml": "yaml",
    }

    return filetypes.get(ext, "")


def matches_glob_pattern(file_path: str, pattern: str) -> bool:
    """Check if a file matches a glob pattern."""
    filename = os.path.basename(file_path)

    # Convert glob pattern to regex
    regex_pattern = wildcard_to_regex(pattern)
    regex = re.compile(regex_pattern)

    # Check both filename and full path
    return bool(regex.search(filename) or regex.search(file_path))


def should_include(
    file_path: str, glob_patterns: list[str], ignore_glob_patterns: list[str]
) -> bool:
    """Check if a file should be included based on glob patterns."""
    # If ignore patterns match, exclude the file
    for pattern in ignore_glob_patterns:
        if matches_glob_pattern(file_path, pattern):
            return False

    # If no include patterns specified, include by default (unless ignored above)
    if not glob_patterns:
        return True

    # Check if file matches any include pattern
    for pattern in glob_patterns:
        if matches_glob_pattern(file_path, pattern):
            return True

    return False


def get_real_path(path: str) -> str:
    """Get the real absolute path using Python's built-in path resolution."""
    try:
        return str(Path(path).resolve())
    except Exception:
        return os.path.abspath(path)


def should_ignore(
    file_path: str,
    ignore_glob_patterns: list[str],
    ignore_dirs: list[str],
) -> bool:
    """Check if a file matches any ignore pattern or is in an ignored directory."""
    real_file_path = get_real_path(file_path)

    for ignore_dir in ignore_dirs:
        if os.sep not in ignore_dir and ignore_dir in file_path.split(os.sep):
            return True

        if os.path.dirname(file_path).endswith(os.sep + ignore_dir):
            return True

        if os.sep in ignore_dir:
            real_ignore_dir = get_real_path(ignore_dir.rstrip("/"))
            if real_file_path.startswith(real_ignore_dir):
                return True

            dir_path = os.path.dirname(file_path)
            if ignore_dir.rstrip("/") in dir_path:
                return True

    for pattern in ignore_glob_patterns:
        if matches_glob_pattern(file_path, pattern):
            return True

    return False


def parse_args() -> Args:
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(
        description="List files and their contents",
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )

    _ = parser.add_argument(
        "-a",
        "--all",
        action="store_true",
        help="Include hidden files",
    )
    _ = parser.add_argument(
        "-r",
        "--recursive",
        action="store_true",
        help="Recursively list files in subdirectories",
    )
    _ = parser.add_argument(
        "--ignore-dir",
        action="append",
        help="Ignore directory DIR (can be used multiple times)",
    )
    _ = parser.add_argument(
        "--globs",
        action="append",
        help="Only include files matching glob pattern (can be used multiple times)",
    )
    _ = parser.add_argument(
        "--ignore-globs",
        action="append",
        help="Ignore files matching glob pattern (can be used multiple times)",
    )
    _ = parser.add_argument(
        "--pattern",
        help="Only show lines matching glob PATTERN",
    )
    _ = parser.add_argument(
        "-n",
        "--line-numbers",
        action="store_true",
        help="Show line numbers",
    )
    _ = parser.add_argument(
        "--debug",
        action="store_true",
        help="Enable debug output",
    )
    _ = parser.add_argument(
        "--omit-bins",
        action="store_true",
        help="Skip binary files in output",
    )
    _ = parser.add_argument(
        "directory",
        nargs="?",
        default=".",
        help="Directory to scan",
    )
    _ = parser.add_argument(
        "files",
        nargs="*",
        help="Specific files to include",
    )

    parsed_args = parser.parse_args()

    args = Args()
    args.show_all = cast(
        bool,
        parsed_args.all,
    )
    args.recursive = cast(
        bool,
        parsed_args.recursive,
    )
    args.debug = cast(
        bool,
        parsed_args.debug,
    )
    args.directory = cast(str, parsed_args.directory)
    args.files = cast(list[str], parsed_args.files or [])
    args.content_pattern = cast(str, parsed_args.pattern or "")
    args.show_line_numbers = cast(bool, parsed_args.line_numbers)
    args.omit_bins = cast(bool, parsed_args.omit_bins)

    ignore_dir_list = cast(list[str] | None, parsed_args.ignore_dir)
    if ignore_dir_list:
        args.ignore_dir.extend(ignore_dir_list)

    globs_list = cast(list[str] | None, parsed_args.globs)
    if globs_list:
        args.globs.extend(globs_list)

    ignore_globs_list = cast(list[str] | None, parsed_args.ignore_globs)
    if ignore_globs_list:
        args.ignore_globs.extend(ignore_globs_list)

    # Process files
    for file in args.files:
        if os.path.isfile(file):
            basename = os.path.basename(file)
            args.globs.append(basename)
        else:
            args.globs.append(file)

    return args


def find_files(args: Args) -> list[str]:
    """Find all files in the directory based on the provided arguments."""
    files: list[str] = []
    maxdepth = float("inf") if args.recursive else 1

    # Walk directory structure
    dir_stack: list[tuple[str, int]] = [(args.directory, 0)]
    while dir_stack:
        current_dir, current_depth = dir_stack.pop()

        if current_depth >= maxdepth:
            continue

        try:
            entries = sorted(os.listdir(current_dir))

            for entry in entries:
                if entry in (".", ".."):
                    continue

                if not args.show_all and entry.startswith("."):
                    continue

                full_path = os.path.join(current_dir, entry)

                if os.path.isdir(full_path):
                    if not should_ignore(
                        full_path,
                        args.default_ignore_globs + args.ignore_globs,
                        args.ignore_dir,
                    ):
                        dir_stack.append(
                            (
                                full_path,
                                current_depth + 1,
                            )
                        )
                    elif args.debug:
                        print(
                            f"Debug: Ignoring directory: {full_path}",
                            file=sys.stderr,
                        )
                elif os.path.isfile(full_path):
                    files.append(full_path)

        except OSError as e:
            print(
                f"Error accessing directory {current_dir}: {e}",
                file=sys.stderr,
            )

    files.sort()
    return files


def process_file(file_path: str, args: Args) -> None:
    """Process a single file and output its contents."""
    if args.directory == ".":
        rel_path: str = file_path
    else:
        try:
            rel_path = str(Path(file_path).relative_to(Path(args.directory)))
        except ValueError:
            rel_path = file_path

    if not should_include(rel_path, args.globs, args.ignore_globs):
        return

    if should_ignore(
        rel_path,
        args.default_ignore_globs + args.ignore_globs,
        args.ignore_dir,
    ):
        if args.debug:
            print(
                f"Debug: Ignoring file: {rel_path}",
                file=sys.stderr,
            )
        return

    if args.omit_bins and is_binary(file_path):
        if args.debug:
            print(
                f"Debug: Skipping binary file: {rel_path}",
                file=sys.stderr,
            )
        return

    safe_path = escape(rel_path)
    print(f'<file path="{safe_path}">')

    if is_binary(file_path):
        print("<binary>true</binary>")  # binary
        print(  # binary content
            "<content>[Binary file - contents not displayed]</content>"
        )
    else:
        filetype = guess_filetype(file_path)
        print(f"<type>{escape(filetype)}</type>")

        try:
            with open(
                file_path,
                "r",
                encoding="utf-8",
                errors="replace",
            ) as f:
                content = f.readlines()

            line_count = len(content)
            filtered_content: list[tuple[int, str]] = []

            if args.content_pattern:
                try:
                    regex_pattern = wildcard_to_regex(args.content_pattern)
                    pattern = re.compile(regex_pattern)
                    for i, line in enumerate(content):
                        if pattern.search(line):
                            filtered_content.append((i + 1, line))
                except re.error as e:
                    print(f"<error>Error in pattern: {escape(str(e))}</error>")
                    filtered_content = [(i + 1, line) for i, line in enumerate(content)]
            else:
                filtered_content = [(i + 1, line) for i, line in enumerate(content)]

            filtered_count = len(filtered_content)

            if filtered_count > 1000 and not args.content_pattern:
                to_display = filtered_content[:100]
                print_trailing_message = True
            else:
                to_display = filtered_content
                print_trailing_message = False

            print("<content>")
            for line_num, line in to_display:
                if args.show_line_numbers:
                    print(
                        f"{line_num:>4}| {line}",
                        end="",
                    )
                else:
                    print(line, end="")

            if print_trailing_message:
                print(f"... ({line_count - 100} more lines)")

            print("</content>")

        except Exception as e:
            print(f"<error>{escape(str(e))}</error>")

    print("</file>")


def main():
    """Main function."""
    args = parse_args()

    if args.debug:
        os.environ["CATLS_DEBUG"] = "1"

    if not os.path.isdir(args.directory):
        print(
            f"Error: '{args.directory}' is not a valid directory.",
            file=sys.stderr,
        )
        sys.exit(1)

    # Clean up ignore directories
    args.ignore_dir = [d.rstrip("/") for d in args.ignore_dir]

    if args.debug:
        print(
            f"Debug: Ignoring directories: {args.ignore_dir}",
            file=sys.stderr,
        )

    # Find all files
    files = find_files(args)

    if not files:
        print(f"No files found in directory: {args.directory}")
        return

    print("<files>")

    for file_path in files:
        process_file(file_path, args)

    print("</files>")


if __name__ == "__main__":
    main()
